src/OrleansRuntime/Streams/PubSub/PubSubRendezvousGrain.cs

```diff
@@ -160,73 +160,80 @@ private static bool IsDeadProducer(IStreamProducerExtension producer)
-                                    bool removed = State.Producers.Remove(producer);
-                                    someProducersRemoved = true; // Re-save state changes at end
-                                    counterProducersRemoved.Increment();
-                                    counterProducersTotal.DecrementBy(removed ? 1 : 0);
-
-                                    // And ignore this error
-                                }
-                                else
-                                {
-                                    throw exc;
-                                }
-                            }
-                        }, TaskContinuationOptions.ExecuteSynchronously);
-                    tasks.Add(addSubscriberPromise);
+                    // Producer is not active (could be stopping / shutting down) so skip
+                    if (logger.IsVerbose) logger.Verbose("Producer {0} on stream {1} is not active - skipping.", producer, streamId);
+                    continue;
                 }
 
-                Exception exception = null;
-                try
-                {
-                    await Task.WhenAll(tasks);
-                }
-                catch (Exception exc)
+                tasks.Add(NotifyProducer(producer, subscriptionId, streamId, streamConsumer, filter));
+            }
+
+            Exception exception = null;
+            try
+            {
+                await Task.WhenAll(tasks);
+            }
+            catch (Exception exc)
+            {
+                exception = exc;
+            }
+
+            // if the number of producers has been changed, resave state.
+            if (State.Producers.Count != initialProducerCount)
+                await WriteStateAsync();
+
+            if (exception != null)
+                throw exception;
+        }
+
+        private async Task NotifyProducer(PubSubPublisherState producer, GuidId subscriptionId, StreamId streamId,
+            IStreamConsumerExtension streamConsumer, IStreamFilterPredicateWrapper filter)
+        {
+            try
+            {
+                await producer.Producer.AddSubscriber(subscriptionId, streamId, streamConsumer, filter);
+            }
+            catch (GrainExtensionNotInstalledException)
+            {
+                RemoveProducer(producer);
+            }
+            catch (OrleansException orleansException)
+            {
+                if (!orleansException.IsClientNotAvailableException())
                 {
-                    exception = exc;
+                    throw;
                 }
+                RemoveProducer(producer);
+            }
+        }
 
-                if (someProducersRemoved)
-                    await WriteStateAsync();
+        private void RemoveProducer(PubSubPublisherState producer)
+        {
+            logger.Warn((int)ErrorCode.Stream_ProducerIsDead,
+                "Producer {0} on stream {1} is no longer active - discarding.",
```

Comment:
Write instead: "Producer {0} on stream {1} is no longer active - permanently removing from the pub sub table.",